### DISCUSSION TITLE
rabbitmq streams compression

### DISCUSSION THREAD

**Discussion by wrobell** [rabbitmq/rabbitmq-server#3996] [Other]
Is it possible to compress stream data stored by RabbitMQ Streams on a disk?

**Comment by acogoluegnes** [comment]
Yes, it's possible, you need to use "sub-batching", that is squeezing several messages in a regular message entry of the publish frame.

[Structure of a simple message entry](https://github.com/rabbitmq/osiris/blob/98989025d07238b856fa7c9b7f793f7b304b5346/src/osiris_log.erl#L182-L190) (note the very first bit is 0)

[Structure of a sub-batch entry](https://github.com/rabbitmq/osiris/blob/98989025d07238b856fa7c9b7f793f7b304b5346/src/osiris_log.erl#L235-L249) (note the very first bit is 1)

You can use "plain" sub-batch entries, you basically squeeze plain messages in a single "slot". There are also different supported [compression algorithms](https://github.com/rabbitmq/rabbitmq-stream-java-client/blob/main/src/main/java/com/rabbitmq/stream/compression/Compression.java).

This is a pure client-side feature, the broker does not compress and decompress, the client libraries do. This implies that the client library must be able to read sub-batch entries when consuming from a stream, a possibly decompress them (an entry in a chunk can be a sub-batch entry).

The Java client has a [setting for sub-batching](https://rabbitmq.github.io/rabbitmq-stream-java-client/snapshot/htmlsingle/#producer-sub-entry-size-configuration-entry). Note [message de-duplication](https://rabbitmq.github.io/rabbitmq-stream-java-client/snapshot/htmlsingle/#outbound-message-deduplication) is guaranteed when using sub-batching.
